[MRG] Integration of GSoC2015 Gaussian Mixture (first step)

tguillemot opened the pull request:
@xuewei4d has done a huge work for the GM classes.
Nevertheless it seems that #4802 is not integrated yet. I thought it was because of the length of the contribution and the difficulty to measure the efficiency of the new classes with respect to the previous ones.
I propose to integrate this work step by step in several pull requests (400 lines max each time). 

This pull request introduces the new abstract class _MixtureBase with some common GMM attributes and methods. Now, GMM derives from _MixtureBase.

@ogrisel @amueller @agramfort

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -112,7 +114,157 @@ def sample_gaussian(mean, covar, covariance_type='diag', n_samples=1,
+    -------
+    X : array, (n_samples, n_features)
+    """
+    X = check_array(X, dtype=[np.float64, np.float32], ensure_2d=False)
```
why not having `ensure_2d=True, allow_nd=False` ?

TomDLT replied:
also if you check X here, you can remove the check in `GMM._fit`

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -112,7 +114,157 @@ def sample_gaussian(mean, covar, covariance_type='diag', n_samples=1,
+        self.n_features_ = None
+        self.converged_ = False
+
+        if n_init < 1:
```
We usually avoid checking the parameters in `__init__`, since a parameter can also be changed with `set_params`

xuewei4d replied:
That is also old lines, written in 2012. That is removed from my PR.

TomDLT commented on sklearn/mixture/tests/test_mixture_base.py:
```diff
@@ -0,0 +1,94 @@
+                  covariance_type):
+    X = []
+    if covariance_type == 'spherical':
+        for k, (w, m, c) in enumerate(zip(weights, means,
```
you seem no to use the `k`

TomDLT commented on sklearn/mixture/tests/test_mixture_base.py:
```diff
@@ -0,0 +1,94 @@
+            X, Y)
+
+
+RandData = RandomData(rng)
```
Only the class name should start with a Capital

TomDLT commented on sklearn/mixture/tests/test_mixture_base.py:
```diff
@@ -0,0 +1,94 @@
+from ...utils.testing import assert_array_equal
+from ...utils.testing import assert_raise_message
+
+rng = np.random.RandomState(0)
```
[it's really ugly :)](https://github.com/scikit-learn/scikit-learn/pull/6125#discussion-diff-50506337)

TomDLT commented on sklearn/mixture/tests/test_mixture_base.py:
```diff
@@ -0,0 +1,94 @@
+    n_components = RandData.n_components
+    n_features = RandData.n_features
+
+    X_bad_dim = rng.rand(n_samples)
```
we probably don't need to generate 500 random float just to check the dimension

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -112,7 +114,157 @@ def sample_gaussian(mean, covar, covariance_type='diag', n_samples=1,
+                             "Estimation requires at least one run" % n_init)
+
+        if covariance_type not in ['spherical', 'tied', 'diag', 'full']:
+                raise ValueError("Invalid value for 'covariance_type': %s "
```
over indentation?

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -112,7 +114,157 @@ def sample_gaussian(mean, covar, covariance_type='diag', n_samples=1,
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+        """
```
Returns?

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -234,34 +386,35 @@ class GMM(BaseEstimator):
-        if covariance_type not in ['spherical', 'tied', 'diag', 'full']:
-            raise ValueError('Invalid value for covariance_type: %s' %
-                             covariance_type)
+        self.weights_ = np.ones(self.n_components) / self.n_components
```
what if we change `self.n_components` after the init ?

xuewei4d replied:
Obviously, this line is from old ```gmm.py``` written in 2010. I don't know why this line is indicated as a new line.

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -646,8 +765,7 @@ def aic(self, X):
         """
-        return - 2 * self.score(X).sum() + 2 * self._n_parameters()
-
+        return super(GMM, self).aic(X).sum() / X.shape[0]
```
```python
def aic(self, X):
"""[...]"""
return -2 * self.score(X) * X.shape[0] + 2 * self._n_parameters()
```

what is this `.sum()` for? isn't `super(GMM, self).aic(X)` a float?
Also you should not divide `2 * self._n_parameters()` by `X.shape[0]`

xuewei4d commented:
I don't understand why you work on ```gmm.py```, because we decided to use its full name ```gaussian_mixture``` as the file name.

TomDLT commented:
I think the goal is to make the diff readable, and then to do a different PR with a simple file renaming

xuewei4d commented:
Well, the new classes in my PR have huge differences. That's not a small amount of diff.

tguillemot commented:
@TomDLT Thanks for your comments, I will correct all those things fastly.

@xuewei4d Maybe I didn't explain my goal correctly and I'm sorry for that.

Of course your PR has huge differences and obviously I have noticed it. 
I think you've done a really good job with your PR. Nevertheless, it is still not included in scikit-learn.
As you propose a PR of ~4500 lines, I though that it should be really difficult for the reviewers to analyse your code in detail (maybe my supposition is wrong ?).
 
My purpose is to merge all your work. I think the best solution is to merge it step by step. It is what I've proposed with this PR. Of course, it's just a little part of your work but it's easier to analyse it in that way. 
If the reviewers want to merge it, I will propose more PR with more of your code until we totally merge it

If it is not a good solution to merge your work, I will just close the PR.

ogrisel commented:
The problem is to deprecate the old `GMM` class without deprecating the old `VBGMM` and `DPGMM` subclasses for now.

So I would suggest to:

- Rename the old `GMM` class as `_GMMBase` (without changing any line of the code) and make `VBGMM` and `DPGMM` derive from it instead of `GMM`,
- Create new `GMM` class that dire that is in the same file but with a deprecation warning in the `__init__` (and no other overridden method).

This `_GMMBase` class should not be used by any of the new gaussian mixture models. It's just a way to deprecate the old classes on at a time to make the pull request small and easier to review for incremental merge to master.

ogrisel commented:
Then include `@xuewei4d` `GaussianMixture` code in its own file as done in #4802: `sklearn/mixture/dp_gaussian_mixture.py`.

tguillemot commented:
In fact, I thought add the `_GMMBase` class and move the `_MixtureBase` class from `gmm.py` to `base.py` at the next PR but it's easier to do it now.

ogrisel commented on sklearn/mixture/gmm.py:
```diff
@@ -112,7 +114,7 @@ def sample_gaussian(mean, covar, covariance_type='diag', n_samples=1,
 
 
-class GMM(BaseEstimator):
+class _GMMBase(_MixtureBase):
```
This should not inherit from the new `_MixtureBase` class. Just `BaseEstimator` as previously. We do not want to change the implementation and behavior of the old `GMM` class in any way besides issuing a deprecation warning in the init.

tguillemot replied:
In that case, I have to send all the code of `_MixtureBase` and `GaussianMixture` in the same PR. 
To simplify the review, I can propose to split the code in different commit. I don't see a better way of not pushing 1000 lines at once.

ogrisel replied:
yes but just the code for GaussianMixture (not the variational Bayes sub classes).

ogrisel commented:
@tguillemot any news on this PR?

tguillemot commented:
Sorry for being late. I'll send everything before the end of the day.

tguillemot commented:
@agramfort @ogrisel I've push the new version. I'll send a python notebook for the performances in a few moment.

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+    return X
+
+
+class MixtureBase(six.with_metaclass(ABCMeta, DensityMixin, BaseEstimator)):
```
Call it BaseMixure

agramfort replied:
Call it BaseMixure

agramfort commented on examples/mixture/plot_gmm_covariances.py:
```diff
@@ -25,33 +25,41 @@
-print(__doc__)
 
 # Author: Ron Weiss <ronweiss@gmail.com>, Gael Varoquaux
+# Modified by Thierry Guillemot to be compatible with the GaussianMixture class
```
Just add your name. Forget The details

agramfort commented on examples/mixture/plot_gmm_sin.py:
```diff
@@ -78,4 +58,36 @@
+X = np.zeros((n_samples, 2))
+step = 4 * np.pi / n_samples
+
+for i in xrange(X.shape[0]):
```
Py3 no xrange

tguillemot replied:
It is not the classical Py2 `xrange` but the one from `sklearn.externals.six.moves`.
I've used `range` instead.

agramfort replied:
ok thx

agramfort commented on examples/mixture/plot_gmm.py:
```diff
@@ -20,6 +20,9 @@
 regularization properties of the inference algorithm.
 """
+
+# Modified by Thierry Guillemot <thierry.guillemot.work@gmail.com>
```
add your name to author list when there is one or remove it. We don't keep the logs of changes in each file

agramfort commented on examples/mixture/plot_gmm_pdf.py:
```diff
@@ -8,6 +8,8 @@
 matrices.
 """
 
+# Modified by Thierry Guillemot <thierry.guillemot.work@gmail.com>
```
same rmk

agramfort commented on examples/mixture/plot_gmm_sin.py:
```diff
@@ -78,4 +58,36 @@
+# Generate random sample following a sine curve
+np.random.seed(0)
+X = np.zeros((n_samples, 2))
+step = 4 * np.pi / n_samples
```
nitpick but we usually write 4. to explicit float types (with the . )

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+
+    Parameters
+    ----------
+    X : array-like, (n_samples, n_features)
```
shape (n_samples, n_features)

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+        self.n_components = n_components
+        self.tol = tol
+        self.reg_covar = reg_covar
+        self.random_state = random_state
```
respect the order of parameters in assigments

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+        self.n_features = None
+        self.converged_ = False
+        self.n_iter_ = 0
+        self.initialized_ = False
```
init method should only do the assignments. This second block should be done in fit

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+                             "Estimation requires at least one iteration"
+                             % self.max_iter)
+
+        if self.reg_covar < 0:
```
0.

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
```
I know we're not always consistent but remove the =

shape (n_samples, n_features)

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+            resp = np.zeros((n_samples, self.n_components))
+            label = cluster.KMeans(n_clusters=self.n_components, n_init=1,
+                                   random_state=random_state).fit(X).labels_
+            resp[range(n_samples), label] = 1
```
np.arange(n_samples)

agramfort replied:
or even better

resp[:, label] = 1

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+            resp[range(n_samples), label] = 1
+        elif self.init_params == 'random':
+            resp = random_state.rand(X.shape[0], self.n_components)
+            resp = resp / resp.sum(axis=1)[:, np.newaxis]
```
resp /= resp.sum(axis=1)[:, np.newaxis]

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
```
no =

please check in the full diff

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+        The method fit the model `n_init` times and set the parameters with
+        which the model has the largest likelihood or lower bound. Within each
+        trial, the method iterates between E-step and M-step for `max_iter`
+        times until the change of likelihood or lower bound is sless than
```
sless -> less

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
```
2 spaces after .

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+        self
+        """
+        X = _check_X(X, self.n_components)
+        self.n_features = X.shape[1]
```
why do you need to store it as attribute? can you avoid it?

besides as it's a data dependent param it should be n_features_

tguillemot replied:
It's used to check that `X` in functions `score`, `score_sample`, `predict`,... has the same `n_features` than the `X` fitted by `fit`.

ogrisel replied:
If you really need to store a data dependent value on the estimator object during fit, it should either not be public or be marked as a fitted parameter by using a `_` suffix to make it explicit that this is not an hyperparameter of the model. hyperparameter are only set via the constructor and the `set_params` method. See: http://scikit-learn.org/stable/developers/contributing.html#apis-of-scikit-learn-objects

ogrisel replied:
Actually in this case you don't need to store `n_features` as once the model is fitted you can always retrieve it with:

```python
n_features = means_.shape[1]
# do some checks to validate the new data here ...
```

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+
+            if not self.converged_:
+                # compute the log-likelihood of the last m-step
+                warnings.warn('Initialization %d is not converged. '
```
did not converge.

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+                best_params = self._get_parameters()
+
+        self._set_parameters(best_params)
+        self.covars_ = self.covariances_
```
don't we rename covars_ to covariances_ ?

tguillemot replied:
Sorry, I've forgotten to remove this line.

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+
+from scipy import linalg
+
+from sklearn.externals.six.moves import zip
```
use relative import

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+    weights : array-like, shape = (n_components,)
+        The proportions of components of each mixture.
+
+    weights_shape = (n_components,)
```
weights_shape: array, shape (n_components,)

or tuple?

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape = (n_samples, n_components)
```
responsibilities -> resp

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape = (n_samples, n_components)
```
responsibilities -> resp

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+        except linalg.LinAlgError:
+            raise ValueError("'covariances' must be symmetric, "
+                             "positive-definite")
+        cv_log_det = 2. * np.sum(np.log(np.diagonal(cov_chol)))
```
use the fast_logdet function we have in extmath.py

agramfort replied:
forget it you use below cov_chol

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+        cv_sol = linalg.solve_triangular(cov_chol, (X - mu).T,
+                                         lower=True).T
+        log_prob[:, k] = np.sum(np.square(cv_sol), axis=1)
+    log_prob = - .5 * (n_features * np.log(2. * np.pi) + cv_log_det + log_prob)
```
this function looks very similar with the one above. Any chance to DRY ?

tguillemot replied:
I've try to find another way to do that. Unfortunately, this code is optimize for each covariance type.
The best I can propose is on that [GIST](https://gist.github.com/tguillemot/2374805a5251f08e630c#file-log_prob-py) but I don't like this solution, it increases a lot the code complexity.

agramfort replied:
ok fair enough

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+
+    reg_covar : float, defaults to 0.
+        Non-negative regularization added to the diagonal of covariance.
+        Allows to assure the positivness of the covariance matrices.
```
Allows to assure that the covariance matrices are all positive.

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+        The number of EM iterations to perform.
+
+    n_init : int, defaults to 1.
+        The number of initializations to perform. The best results is kept.
```
The number of K-means initializations to perform

agramfort replied:
forget it.

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+        True when convergence was reached in fit(), False otherwise.
+
+    initialized_ : bool
+        True when fit() has been apply at least once, False otherwise.
```
applied

agramfort replied:
attributes with trailing _ should not exist before a call to fit

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+
+        Returns
+        -------
+        bic: float (the greater the better)
```
put "the greater the better" on the next line.

agramfort commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+    covars_bad = rng.rand(n_components + 1, n_features)
+    g = GaussianMixture(n_components=n_components, covariances_init=covars_bad,
+                        covariance_type='diag')
+    g.n_features = n_features
```
I don't like these.

agramfort commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+    n_samples, n_features, n_components = 50, 3, 2
+    X = rng.randn(n_samples, n_features)
+    # standard gaussian entropy
+    SGH = 0.5 * (np.log(np.linalg.det(np.cov(X.T, bias=1))) +
```
don't use capital letters

agramfort commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+        g = GaussianMixture(
+            n_components=n_components, covariance_type=cv_type,
+            random_state=rng)
+        g.n_features = n_features
```
you should not have to do this

agramfort commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+
+    assert_almost_equal(g.weights_, h.weights_)
+    assert_almost_equal(g.means_, h.means_)
+    assert_almost_equal(g.covariances_, h.covariances_)
```
regarding tests 

lines 398, 424 are not covered.

I don't see any test of score and score_samples methods

agramfort commented:
that's it for me

nice job @tguillemot !

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+                # compute the log-likelihood of the last m-step
+                warnings.warn('Initialization %d is not converged. '
+                              'Try different init parameters, '
+                              'or increase n_init, tol '
```
User should check for `max_iter` rather than `n_init`: we are inside the `n_init` loop here.

TomDLT commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,673 @@
+        return np.log(self.weights_)
+
+    def _check_is_fitted(self):
+        check_is_fitted(self, 'weights_')
```
you can give a list `['weights_', 'means_', 'covariances']` to have only one call to `check_is_fitted`

TomDLT commented on sklearn/mixture/gmm.py:
```diff
@@ -649,6 +649,23 @@ def aic(self, X):
+            n_iter=n_iter, n_init=n_init, params=params,
+            init_params=init_params, verbose=verbose)
+
+        warnings.warn(
```
Isn't it redundant with the `@deprecated` decorator ?
If not, should we use  `warnings.warn(msg, category=DeprecationWarning)`?

TomDLT commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+                         "Invalid value for 'n_init': %d "
+                         "Estimation requires at least one run"
+                         % n_init,
+                         gmm._check_initial_parameters)
```
You shouldn't call `_check_initial_parameters` here.
You never use this function, yet your test is not able to tell you that.

TomDLT commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,425 @@
+        """Print verbose message on initialization."""
+        if self.verbose == 1:
+            print("Initialization %d" % n_init)
+        elif self.verbose == 2:
```
You should probably use `self.verbose >= 2`, as it is tempting to set `verbose=10` to have full verbosity without checking what is the highest value used in the code.

TomDLT commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+            sys.stdout = old_stdout
+
+
+def test_warm_start():
```
This test actually don't test anything, as you can see by changing `max_iter=0` in `h`.
If you plot the data (see below), you can realize that the clusters are very far from each other (the `self.means` are too large in `RandomData`), which makes the convergence complete just after K-means initialization.

```python
import matplotlib.pyplot as plt
plt.plot(X[:, 0], X[:, 1], 'o')
plt.show()
```
You should first reduce the distance between clusters, to have some overlapping.
Also to be sure this test is useful, you should design a failing test, that does not fail with warm start.

TomDLT commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+        try:
+            g.fit(X)
+            h.fit(X)
+            k.fit(X)
```
what are you testing here?
This test should reveal that `verbose==20` does not print anything.

TomDLT commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,681 @@
+        bic = (2 * n_samples * SGH +
+               np.log(n_samples) * g._n_parameters())
+        bound = n_features * 3. / np.sqrt(n_samples)
+        assert_true(np.abs(g.aic(X) - aic) / n_samples < bound)
```
How do you come up with this bound?

tguillemot replied:
This bound is also defined in the old `test_gmm.py`, I thought it was a know property of BIC and AIC.
@bthirion Do you know where `bound` comes from ?

bthirion replied:
On 21/03/2016 13:32, tguillemot wrote:
>
> In sklearn/mixture/tests/test_gaussian_mixture.py 
> <https://github.com/scikit-learn/scikit-learn/pull/6407#discussion_r56814127>:
>
> > +    rng = np.random.RandomState(0)
> > +    n_samples, n_features, n_components = 50, 3, 2
> > +    X = rng.randn(n_samples, n_features)
> > +    # standard gaussian entropy
> > +    SGH = 0.5 * (np.log(np.linalg.det(np.cov(X.T, bias=1))) +
> > +                 n_features * (1 + np.log(2 * np.pi)))
> > +    for cv_type in COVARIANCE_TYPE:
> > +        g = GaussianMixture(
> > +            n_components=n_components, covariance_type=cv_type,
> > +            random_state=rng)
> > +        g.fit(X)
> > +        aic = 2 * n_samples * SGH + 2 * g._n_parameters()
> > +        bic = (2 * n_samples * SGH +
> > +               np.log(n_samples) * g._n_parameters())
> > +        bound = n_features * 3. / np.sqrt(n_samples)
> > +        assert_true(np.abs(g.aic(X) - aic) / n_samples < bound)
>
> This bound is also defined in the old |test_gmm.py|, I thought it was 
> a know property of BIC and AIC.
> @bthirion <https://github.com/bthirion> Do you know where |bound| 
> comes from ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/scikit-learn/scikit-learn/pull/6407/files/392e3d164de5cd66f43a297c282142f22654c80c#r56814127>
>
If I understand correctly, it is just an estimate of the expected 
varaibility about the theoretical aic value. By the law of large numbers 
it should not be more extreme than O(1 / sqrt(n_samples)). Taking 3. / 
sqrt(n_samples) is however not a very afe choice (you expect more 
extreme value by chance in ~ 1/1000 of the cases).
HTH

tguillemot replied:
Thank you for the answer. 
I've changed the test for 1 / sqrt(n_samples).

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -75,21 +75,21 @@ def __init__(self, n_components, tol, reg_covar,
-        """Check values of the basic parameters."""
+        Parameters
+        ----------
+        X : array-like, shape  (n_samples, n_features)
```
2 spaces after shape

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -128,10 +133,10 @@ def _initialize_parameters(self, X):
             label = cluster.KMeans(n_clusters=self.n_components, n_init=1,
                                    random_state=random_state).fit(X).labels_
-            resp[range(n_samples), label] = 1
+            resp[np.arange(n_samples), label] = 1
```
resp[:, label] = 1

tguillemot replied:
I'm not agree with you. `resp[:, label]` will select the set :
```
[[ resp[0, label[0]], resp[0, label[1]], ... resp[0, label[n]] ],
 [ resp[1, label[0]], resp[1, label[1]], ... resp[1, label[n]] ],
 ...
 [ resp[n, label[0]], resp[n, label[1]], ... resp[n, label[n]] ]]
```
not the set :
```
[ resp[0, label[0]], resp[1, label[1]], ..., resp[n, label[n] ]
```
To select the second set, the only way is to do `resp[np.arange(n_samples), label]`

agramfort replied:
ok got it ! sorry for the noise

agramfort commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -644,38 +615,90 @@ def test_gaussian_mixture_verbose():
+    # Check score value
+    gmm1.fit(X)
+    gmm_score = gmm1.score(X)
+    gmm_score_proba = gmm1.score_samples(X).mean()
```
can you check that gmm1.score_samples(X).shape[0] == n_samples?

tguillemot replied:
Ok. I prefer to put this test with `test_score_samples`.

agramfort commented:
travis is not happy.

that's it for me !

thx @tguillemot

tguillemot commented:
I've upload a [notebook](https://github.com/tguillemot/GSoC-Gaussian-Mixture-Analysis/blob/master/gaussian_mixture_benchmark.ipynb) to show the performances of the new GaussianMixture class and a comparison with the old GMM class.

tguillemot commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+                             % self.covariance_type)
+
+        param_shape = self._define_parameter_shape(int(X.shape[1]))
+
```
To solve the Travis problem (with Python-2.7 -- x64), I have put `int(X.shape[1])`.
I don't like it but I haven't find a better solution.

TomDLT replied:
That means that we can't use more than `2 ** 31 = 2.1e9` features.

As `assert_raise_message` does not need the full message, I would rather change the failing test into:
```python
msg = "The parameter 'means' should have the shape of "
assert_raise_message(ValueError, msg, g.fit, X)
```
which does not weaken so much the test.
___
Appveyor's problem: https://github.com/numpy/numpy/issues/5809

tguillemot replied:
Thanks @TomDLT , I've corrected it.

tguillemot commented:
@agramfort @ogrisel @TomDLT Do you have other comments ? Are you OK to merge ?

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -139,7 +139,8 @@ VBGMM: variational Gaussian mixtures
-model with :ref:`variational inference <variational_inference>` algorithms. The API is identical to
+model with :ref:`variational inference <variational_inference>` algorithms.
+The API is identical to
 :class:`GMM`. It is essentially a middle-ground between :class:`GMM`
```
Should we remove all the references to `:class:'GMM'`?

tguillemot replied:
As VBGMM and DPGMM are not changed yet, I hesitated to remove these lines.
Maybe, it's better to remove it definitely.

agramfort replied:
get rid of all references to deprecated code

tguillemot commented:
#6305 porposes to add GMM to plot_cluster_comparison.py. I think, it could be a good idea to add it with this PR. What do you think ?

bthirion commented:
I would keep this for a future PR.

tguillemot commented:
OK, thanks @bthirion.

agramfort commented on examples/mixture/plot_gmm.py:
```diff
@@ -66,16 +50,36 @@
+X = np.r_[np.dot(np.random.randn(n_samples, 2), C),
+          .7 * np.random.randn(n_samples, 2) + np.array([-6, 3])]
+
+# Fit a mixture of Gaussians with EM using five components
```
mixture of Gaussians -> Gaussian mixture

agramfort commented on examples/mixture/plot_gmm.py:
```diff
@@ -66,16 +50,36 @@
+plot_results(X, gmm.predict(X), gmm.means_, gmm.covariances_, 0,
+             'Gaussian Mixture')
+
+# Fit a Dirichlet process mixture of Gaussians using five components
```
mixture of Gaussians -> Gaussian mixture

git grep to see this elsewhere

agramfort commented on examples/mixture/plot_gmm_selection.py:
```diff
@@ -14,17 +14,20 @@
-print(__doc__)
 
-import itertools
+# Modified by Thierry Guillemot <thierry.guillemot.work@gmail.com>
```
remove

agramfort commented on examples/mixture/plot_gmm_selection.py:
```diff
@@ -41,7 +44,8 @@
 for cv_type in cv_types:
     for n_components in n_components_range:
         # Fit a mixture of Gaussians with EM
```
Gaussian Mixture

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,427 @@
+
+            if do_init:
+                self._initialize_parameters(X)
+                self.initialized_ = True
```
put this line in _initialize_parameters method

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,427 @@
+            raise ValueError("Unimplemented initialization method '%s'"
+                             % self.init_params)
+
+        self._initialize(X, resp)
```
here

self.initialized_ = True

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,427 @@
+
+        Returns
+        -------
+        labels : array, shape (n_samples,) component labels
```
don't write param desc on this line.

also there is no description of X above

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+
+    n_components : int
+
+    n_features : int
```
bad docstring

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+
+    Returns
+    -------
+    Sk : array, shape (n_components, n_features)
```
Sk -> covariances

agramfort replied:
and think shape is wrong

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+
+    Returns
+    -------
+    Sk : array, shape (n_components, n_features)
```
same rmk

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+    """
+    avg_X2 = np.dot(resp.T, X * X) / nk[:, np.newaxis]
+    avg_means2 = means ** 2
+    avg_X_means = means * np.dot(resp.T, X) / nk[:, np.newaxis]
```
probably not critical but some of these operations can be done inplace

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+        The centers of the current components.
+
+    covariances : array
+        The sample covariances of the current components.
```
explain that shapes can be different based on type

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+
+    initialized_ : bool
+        True when fit() has been applied at least once, False otherwise.
+        `initialized_` will not exist before a call to fit.
```
in the end is it necessary to have initialized_ param? as for example covariances_ will not exist if fit was not called. So you already know if it was initialized_ with other attributes no?

tguillemot replied:
It's true indeed. I remove initialized_.

agramfort commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,677 @@
+
+        Parameters
+        ----------
+        X : array of shape(n_samples, n_dimensions)
```
space after shape

tguillemot commented:
Thanks @agramfort. I've pushed the correction.

ogrisel commented:
@tguillemot could you please run some benchmarks to check convergence time, convergence quality and memory usage (using `memory_profiler`) of the new GaussianMixture class vs the GMM class in master?

You can make 2 or 3 synthetic datasets with `sklearn.datasets.make_blobs`, for instance one with 1000 samples and 10 features and another one with 1e5 samples and 50 features.

And run benchmarks for various numbers of components and covariance matrix types.

ogrisel commented:
Oops sorry I had not seen the notebook. Great work, I will have a look in more details.

tguillemot commented:
A week ago, pomegranate has done a comparison of the efficiency between the old GMM class and their promegranate.GeneralMixtureModel. You can find it here : [promegranate originial benchmark](https://github.com/jmschrei/pomegranate/blob/master/benchmarks/pomegranate_vs_sklearn_gmm.ipynb).
Nevertheless, I've preferred check their work and compare the performances with the new GaussianMixture class. The results are here : [pomegranate vs GaussianMixture results](https://github.com/tguillemot/promegranate_benchmark/blob/master/pomegranate_vs_sklearn_gmm.ipynb).
I've put some comment in bold to explain what I've changed between the two versions.

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,440 @@
+                # since we compute the log-likelihood of the initialization
+                max_log_likelihood = current_log_likelihood
+                best_params = self._get_parameters()
+                self.best_n_iter_ = n_iter
```
Please instead store `best_n_iter` as a local variable here.

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,440 @@
+            # compute the log-likelihood of the last m-step
+            current_log_likelihood, _ = self._e_step(X)
+
+        self._set_parameters(best_params)
```
Please store `n_iter_` here (this is the standard attribute name for estimators that accept a `max_iter` parameter in their constructor):

```python
self.n_iter_ = best_n_iter
```

ogrisel replied:
Also the docstring of the class need to be updated to document the `n_iter_` attribute.

ogrisel commented:
I get convergence warnings even when running very simple examples such as:

```
$ python examples/mixture/plot_gmm.py
/Users/ogrisel/code/scikit-learn/sklearn/mixture/base.py:224: ConvergenceWarning: Initialization 1 did not converged. Try different init parameters, or increase n_init, tol or check for degenerate data.
  % (init + 1), ConvergenceWarning)
```

that sounds very fishy to me. Maybe the choice for the default convergence parameters is not adequate?

tguillemot commented:
In fact, between the old and the new GMM class, the `tol` value has changed from 1e-3 to 1e-6.
Moreover the tolerance formula has changed too. 

If `tol_old` is the tol value of the old GMM class, the new tolerance `tol_new` of `GaussianMixture` is defined as :
`tol_new = tol_gmm / n_components`

I think the new definition is better than before but we definitely have to change the default value.
A value of 1e-2 works and seems equivalent to what we had before.
What do you think ?

ogrisel commented:
I have no idea why normalizing the tolerance by the inverse of the number of components is better. Do you have any insight to motivate this re-parametrization (maybe @xuewei4d could share his opinion on this issue)?

Is there an simple empirical way to demonstrate that this is a better parametrization, for instance on a toy dataset?

ogrisel commented:
Left todo:

- [ ] add a test that checks that when `reg_covar=0` we can incrementally fit a model (do one iteration each time by starting with `max_iter=1` and then incrementing `max_iter` each time and calling fit over and over again on the same data with `warm_start=True` and that the likelihood on the training set is always increasing (monotonic).

- [ ] add a test that shows that we get an exception (e.g. `DivergenceError` or something similar) with a meaningful error message that explains that the algorithm has diverged because of a singularity (e.g. too many components) and that adding regularization to the covariance is required.

- [ ] change the default value of `reg_covar` to 1e-6 or something to avoid most common singularity by default.

Right now we get (when editing the example to have `n_components=n_samples`):

```
$ python examples/mixture/plot_gmm.py
Traceback (most recent call last):
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/gaussian_mixture.py", line 313, in _estimate_log_gaussian_prob_full
    cov_chol = linalg.cholesky(cov, lower=True)
  File "/Users/ogrisel/venvs/py35/lib/python3.5/site-packages/scipy/linalg/decomp_cholesky.py", line 81, in cholesky
    check_finite=check_finite)
  File "/Users/ogrisel/venvs/py35/lib/python3.5/site-packages/scipy/linalg/decomp_cholesky.py", line 30, in _cholesky
    raise LinAlgError("%d-th leading minor not positive definite" % info)
numpy.linalg.linalg.LinAlgError: 2-th leading minor not positive definite

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "examples/mixture/plot_gmm.py", line 77, in <module>
    n_components=n_samples, covariance_type='full').fit(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/base.py", line 199, in fit
    current_log_likelihood, resp = self._e_step(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/gaussian_mixture.py", line 568, in _e_step
    log_prob_norm, _, log_resp = self._estimate_log_prob_resp(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/base.py", line 408, in _estimate_log_prob_resp
    weighted_log_prob = self._estimate_weighted_log_prob(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/base.py", line 358, in _estimate_weighted_log_prob
    return self._estimate_log_prob(X) + self._estimate_log_weights()
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/gaussian_mixture.py", line 586, in _estimate_log_prob
    X, self.means_, self.covariances_)
  File "/Users/ogrisel/code/scikit-learn/sklearn/mixture/gaussian_mixture.py", line 315, in _estimate_log_gaussian_prob_full
    raise ValueError("'covariances' must be symmetric, "
ValueError: 'covariances' must be symmetric, positive-definite
```

This error message is not very helpful. We should help the use fix the problem by telling he/her that there are either too many components or that `reg_covar` should be increased.

ogrisel commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,703 @@
+    h = GaussianMixture(n_components=n_components, n_init=1,
+                        max_iter=1, reg_covar=0, random_state=random_state,
+                        warm_start=True)
+    h.fit(X).fit(X)
```
This should not change anything. Instead the following should do one more iteration:

```python
h.fit(X).set_params(max_iter=2).fit(X)
```

TomDLT replied:
Really? We have the same behavior in `LogisticRegression` with warm start:
```python
lr.fit(X, y).fit(X ,y)
# !=
lr.fit(X, y)
```

ogrisel replied:
Indeed you are right, I was mistaken. ElasticNet / Lasso follow the same pattern.

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,441 @@
+
+            current_log_likelihood = -np.infty
+
+            for n_iter in range(self.max_iter):
```
when `warm_start` is `True` and `self.n_iter_` is defined we should only perform `self.max_iter - self.n_iter_` additional iterations to be consistent with the meaning of `max_iter` of other models.

tguillemot commented:
> I have no idea why normalizing the tolerance by the inverse of the number of components is better.

By saying that, I was thinking to the use of the GaussianMixture class. I thought that a lot of users will try different n_components and use the bic value to find the best result. To have a fair comparison between them, you need to have the same tolerance criterium. 
Moreover, classically in the EM algorithm, we use a tolerance which is independent of the number of component. For example, the tolerance value of pomegranate is defined in that way.

Anyway, it's not a big deal, so I can change that if you want.

I've correct your suggestion. I will push them when we are OK with the tolerance.

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,441 @@
+                change = current_log_likelihood - prev_log_likelihood
+                self._print_verbose_msg_iter_end(n_iter, change)
+
+                if abs(change) < self.tol:
```
Actually I don't see where the `tol / n_components` normalization is done. It seems to use the raw tolerance passed by the user here.

tguillemot replied:
It's not the new code but the old code who was doing that. Sorry for not being clear.

ogrisel commented on sklearn/mixture/gaussian_mixture.py:
```diff
@@ -0,0 +1,641 @@
+
+    def _e_step(self, X):
+        log_prob_norm, _, log_resp = self._estimate_log_prob_resp(X)
+        self._log_likelihood = np.sum(log_prob_norm)
```
I would rather not store `_log_likelihood` on self if this is not necessary.

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,441 @@
+        return logsumexp(self._estimate_weighted_log_prob(X), axis=1)
+
+    def score(self, X, y=None):
+        """Compute the log-likelihood of the model on the given data X.
```
The docstring is not correct: this is the per-sample average log-likelihood: as we use `p.mean()` instead of `p.sum()`.

ogrisel commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,441 @@
+            Log likelihood of the Gaussian mixture given X.
+        """
+        p = self.score_samples(X)
+        return p.mean()
```
This should be simplified to:

```python
return self.score_samples(X).mean()
```

ogrisel commented:
> To have a fair comparison between them, you need to have the same tolerance criterium.
Moreover, classically in the EM algorithm, we use a tolerance which is independent of the number of component.

Alright but I don't understand why dividing tol by `n_components` would make stopping criterion (based on a change in the total log likelihood of the training set).

Also shouldn't we divide tol the log-likelihood change by `n_samples` to get a stopping criterion that is less dependent on the size of the training set?

tguillemot commented:
My mistake on the old class it's not divide by `n_components` but by `n_samples` indeed. 
On the GMM the current_log_likelihood is define as mean and not as a sum.
Forget all I've said about the `n_components`. Sorry for the noise.
I change that.

ogrisel commented:
Ok great that makes sense to me then. It would be good to check how other packages check convergence for GMM trained with EM:

http://www.mixmod.org
https://cran.r-project.org/web/views/Cluster.html (bgmm for instance)

ogrisel commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,743 @@
+    for cov_type in COVARIANCE_TYPE:
+        gmm = GaussianMixture(n_components=n_samples, covariance_type=cov_type,
+                              reg_covar=0)
+        assert_raise_message(ValueError,
```
I wonder if we should not introduce a dedicated exception such as `DivergenceError` in `sklearn.exceptions` but maybe we can stick to `ValueError` for now.

tguillemot replied:
I will do another PR for that once we have finished with the GMM, ok ?

ogrisel commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,743 @@
+                              reg_covar=0)
+        assert_raise_message(ValueError,
+                             "The algorithm has diverged because of a "
+                             "singularity (e.g. too many components). "
```
too few samples per component.

ogrisel commented on sklearn/mixture/tests/test_gaussian_mixture.py:
```diff
@@ -0,0 +1,743 @@
+def test_score_samples():
+    random_state = 0
+    cov_type = 'full'
+    rng = np.random.RandomState(random_state)
```
Please directly write:

```python
rng = np.random.RandomState(0)
```

instead of introducing a local variable `random_state` that is used only once. This pattern has been repeated several times in this file.

tguillemot commented:
Ok I need help to solve the travis problem. 
It seems that there is a problem with the `test_base_mixture.py` file which doesn't exist.
More interesting, the test are passing for Python-2.6 but not for the other ?
On my local branch Travis works without problems.

Do you have any explination ?
At the beginning of this branch I have created a file call `test_base_mixture.py` but it was removed a long time ago.

TomDLT commented:
> At the beginning of this branch I have created a file call test_base_mixture.py but it was removed a long time ago.

This is weird. Maybe try to squash and rebase on top of master.

tguillemot commented:
Another solution could be to propose a new PR with a new branch.
I'll do that only if you don't see another solution.

TomDLT commented:
or
```bash
# delete your branch
git checkout GSoC-integration-step1
git checkout -b GSoC-integration-step1_backup
git branch -D GSoC-integration-step1
# create new branch with same name
git checkout master
git checkout -b GSoC-integration-step1
# cherry-pick and push
git cherry-pick 628e6a56c861aa582cf7f6d4032c9bff24bd07de
git push -f origin GSoC-integration-step1
git branch -D GSoC-integration-step1_backup
```

TomDLT commented:
Maybe do some make clean also, it might help

tguillemot committed:
Integration of the new GaussianMixture class.

Depreciation of the GMM class.

Modification of the GaussianMixture class.
  Some functions from the original GSoC code have been removed, renamed or simplified.
  Some new functions have been introduced (as the 'check_parameters' function).
  Some parameters names have been changed :
   - covars_ -> covariances_ : to be coherent with sklearn/covariances
  Addition of the parameter 'warm_start' allowing to fit data by using the previous computation.

The old examples have been modified to replace the deprecated GMM class by the new GaussianMixture class.
Every exemple use the eigenvectors norm to solve the scale ellipse problem (Issues 6548).

Correction of all commentaries from the PR
- Rename MixtureBase -> BaseMixture
- Remove n_features_
- Fix some problems
- Add some tests

Correction of the bic/aic test.

Fix the test_check_means and test_check_covariances.

Remove all references to the deprecated GMM class.

Remove initialized_.
Add and correct docstring.

Correct the order of random_state.

Fix small typo.

Some fix in prevision of the integration of the new BayesianGaussianMixture class.

Modification in preparation of the integration of the BayesianGaussianMixture class.
Add 'best_n_iter' attribute.
Fix some bugs and tests.

Change the parameter order in the documentation.

Change best_n_iter_ name to n_iter_.

Fix of the warm_start problem.
Fix the divergence error message.
Correction of the random state init in the test file.
Fix the testing problems.

tguillemot commented:
Ok I've tried your solution and also try to do it manually and it still doesn't work.
Thanks for your solutions @TomDLT .

tguillemot committed:
Add file test_mixture_base.py.

tguillemot committed:
Remove the test_mixture_base.py file.

TomDLT commented:
Interestingly you solved Travis py27 but not Travis py35.
This is nonsense

tguillemot commented:
To solve the problem with the strange travis bug, I've created another PR #6666.
